cdc/sink/mysql.go

```diff
@@ -861,13 +869,53 @@ func (w *mysqlSinkWorker) cleanup() {
 	}
 }
 
-func (s *mysqlSink) Close() error {
+func (s *mysqlSink) Close(ctx context.Context) error {
 	s.execWaitNotifier.Close()
 	s.resolvedNotifier.Close()
 	err := s.db.Close()
 	return cerror.WrapError(cerror.ErrMySQLConnectionError, err)
 }
 
+func (s *mysqlSink) Barrier(ctx context.Context) error {
+	warnDuration := 3 * time.Minute
+	ticker := time.NewTicker(warnDuration)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-ticker.C:
+			ticker.Reset(warnDuration)
+			log.Warn("Barrier doesn't return in time, may be stuck",
+				zap.Uint64("resolved-ts", atomic.LoadUint64(&s.maxResolvedTs)),
+				zap.Uint64("checkpoint-ts", s.checkpointTs()))
+		default:
+			maxResolvedTs := atomic.LoadUint64(&s.maxResolvedTs)
+			if s.checkpointTs() >= maxResolvedTs {
+				return nil
+			}
+			checkpointTs, err := s.FlushRowChangedEvents(ctx, maxResolvedTs)
+			if err != nil {
+				return err
+			}
+			if checkpointTs >= maxResolvedTs {
+				return nil
+			}
```

Comment:
Do we need to add sleep here to avoid hogging CPU?

Reply:
addressed